pyro/distributions/torch_distribution.py

```diff
@@ -221,6 +254,142 @@ def check(self, value):
+        # e.g. (2, 4) -> (2, 1)
+        proposed_shape = broadcast_shape(self.batch_shape, batch_shape)
+        if tuple(reversed(proposed_shape)) > tuple(reversed(batch_shape)):
+            raise ValueError("Existing batch shape {} cannot be expanded "
+                             "to the new batch shape {}."
+                             .format(self.batch_shape, batch_shape))
+        # Adjust existing sample shape if possible.
+        base_dist = self.base_dist
+        base_batch_shape = batch_shape + self.event_shape[:self.reinterpreted_batch_ndims]
+        cut = len(base_batch_shape) - len(base_dist.batch_shape)
+        left, right = base_batch_shape[:cut], base_batch_shape[cut:]
+        if right == base_dist.batch_shape:
+            sample_shape = left
+        # Modify the base distribution's batch shape,
+        # if existing sample shape cannot be adjusted.
+        else:
+            base_dist = self.base_dist.expand(base_batch_shape)
+            assert not isinstance(base_dist, ReshapedDistribution)
+            sample_shape = torch.Size(())
+        return ReshapedDistribution(base_dist, sample_shape, self.reinterpreted_batch_ndims)
+
+    def expand_by(self, sample_shape):
+        base_dist = self.base_dist
+        sample_shape = torch.Size(sample_shape) + self.sample_shape
+        reinterpreted_batch_ndims = self.reinterpreted_batch_ndims
+        return ReshapedDistribution(base_dist, sample_shape, reinterpreted_batch_ndims)
+
+    def independent(self, reinterpreted_batch_ndims=None):
+        if reinterpreted_batch_ndims is None:
+            reinterpreted_batch_ndims = len(self.batch_shape)
+        base_dist = self.base_dist
+        sample_shape = self.sample_shape
+        reinterpreted_batch_ndims = self.reinterpreted_batch_ndims + reinterpreted_batch_ndims
+        return ReshapedDistribution(base_dist, sample_shape, reinterpreted_batch_ndims)
+
+    @property
+    def has_rsample(self):
+        return self.base_dist.has_rsample
+
+    @property
+    def has_enumerate_support(self):
+        return self.base_dist.has_enumerate_support
+
+    @constraints.dependent_property
+    def support(self):
+        return IndependentConstraint(self.base_dist.support, self.reinterpreted_batch_ndims)
+
+    @property
+    def _validate_args(self):
+        return self.base_dist._validate_args
+
+    @_validate_args.setter
+    def _validate_args(self, value):
+        self.base_dist._validate_args = value
+
+    def sample(self, sample_shape=torch.Size()):
+        return self.base_dist.sample(sample_shape + self.sample_shape)
+
+    def rsample(self, sample_shape=torch.Size()):
+        return self.base_dist.rsample(sample_shape + self.sample_shape)
+
+    def log_prob(self, value):
+        shape = broadcast_shape(self.batch_shape, value.shape[:value.dim() - self.event_dim])
+        return sum_rightmost(self.base_dist.log_prob(value), self.reinterpreted_batch_ndims).expand(shape)
```

Comment:
This should be the way to do! Expand parameters first, then computing log_prob might be ineffective if log_prob contains expensive operators.

cc @fritzo for a throughout discussion about this problem

Reply:
At a primary glance, it doesn't seem like we should be introducing the ReshapedDistribution back, but instead patch the mvn and lowrank mvn distributions, as you have done below. I'll gather a few test statistics and close this PR.